src/content/docs/concepts/widgets.md

```diff
@@ -0,0 +1,308 @@
+        let greeting = format!("Hello, {}!", self.name);
+        buf.set_string(area.x, area.y, greeting, Style::default());
+    }
+}
+```
+
+In this example, the `GreetingWidget` struct has a single field, `name`, which is a `String`. The
+`render` function takes a `Rect` and a mutable reference to a `Buffer`, and sets the string at the
+specified coordinates in the buffer.
+
+Widgets are not restricted to just calling methods on `Buffer`. They can also create and render
+other widgets within their `render` method. For example, instead of directly calling methods on
+`buf`, a Widget can create a `Line` widget with a vector of `Span`s, where the `Span` for the name
+is style. This `Line` widget can then be rendered within the Widget's `render` method.
+
+Here's the same greeting example using nested widgets:
+
+```rust
+struct GreetingWidget {
+    name: String,
+}
+
+impl Widget for GreetingWidget {
+    fn render(self, area: Rect, buf: &mut Buffer) {
+        let hello = Span::raw("Hello, ");
+        let name = Span::styled(self.name, Modifier::BOLD);
+        let line = Line::from(vec![hello, name]);
+        line.render(area, buf);
+    }
+}
+```
+
+This approach allows for composing and and reusing widgets. For example, the `Line` widget can be
+used in other widgets or even in other parts of the application. Additionally, it allows for easier
+testing and maintenance of widgets as the code related to rendering is organized in a consistent
+place (the `impl Widget` blocks).
+
+# Implementing Stateful Widgets
+
+In some situations, a widget might need to be able to mutate some extra state while rendering
+itself. An example of this is how the built-in `List` widget works. During rendering the `List`
+updates the scroll position in the state to ensure that the selected item is visible in the
+rendering area.
+
+Here's an example of implementing the `StatefulWidget` trait for a frame count Widget:
+
+```rust
+struct FrameCountWidget {
+    style: Style,
+}
+
+impl StatefulWidget for CounterWidget {
+    type State = i32;
+
+    fn render(self, area: Rect, buf: &mut Buffer, state: &mut i32) {
+        *state += 1;
+        let text = format!("Frame count: {state}");
+        Line::styled(text, style).render(area, buf);
+    }
+}
+```
+
+In this example, the `FrameCount` widget increments the state every time it is rendered, counting
+the number of frames.
+
+# Implementing WidgetRef
+
+In Ratatui 0.26.0, we added the [`WidgetRef`] (and similarly [`StatefulWidgetRef`]) traits. These
+allow widgets to be created that are rendering by reference, which can be useful for storing a
+widget which can be rendered multiple times instead of constructing it on every frame. This also
+makes it easy to create dynamic collections of widgets (e.g. Panes in a layout) by boxing the
+widgets using `Box<dyn T>`.
+
+Note: this requires the `unstable-widget-ref` feature flag to be enabled (though this should be
+stabilized soon).
```

Comment:
This will make it an highlighted note:

```suggestion
::: note

This requires the `unstable-widget-ref` feature flag to be enabled (though this should be
stabilized soon).

:::
```